File: test_dir/python/src/tuples/first.py
from test_dir.python.pypp.ownership import Valu
from test_dir.python.pypp.tuple_get import tg


def _inline_tuple(tup: Valu(tuple[float, str])):
    print(tup)


def _get_tup() -> tuple[int, float]:
    return 1, 2.0


def _argument_unpacking(a: int, b: float):
    print(a, b)


def _arg_unpacking_fail(a: int, b: int, c: int):
    print(a, b, c)


def tuples_fn():
    print("TUPLE RESULTS:")
    a: tuple[int, float, str] = (1, 1.2, "a")
    # count
    print(str(a.count(2)))
    # index
    print(str(a.index(1.2)))
    # print(a.index(1.1))  # Throws error
    # access
    b: int = tg(a, 0)
    print(str(b))
    # comparisons
    print(str((1, 2) == (1, 2)))
    print(str((1, 2) != (1, 2)))
    print(str((1, 2) < (1, 2)))
    print(str((1, 2) <= (1, 2)))
    print(str((1, 2) > (1, 2)))
    print(str((1, 2) >= (1, 2)))
    # printing
    print((1, 2))
    print((1, 2, "a"))
    # length
    print(str(len((1, 2))))
    # inline passing
    _inline_tuple((1.2, "z"))
    # unpacking elements
    x, y, z = a
    print(x, y, z)
    # unpacking elements from a function
    u, v = _get_tup()
    print(u, v)
    # argument unpacking
    _argument_unpacking(*_get_tup())
    # argument unpacking only works if it is the only argument to the call.
    # arg_unpacking_fail(1, *get_tup())

    # variables that are passed to tuples are moved
    c: list[int] = [1, 2, 3]
    d: tuple[int, list[int]] = (1, c)
    print(d)
    print(
        "below will be [1, 2, 3] for Python, but [] for C++ because the list was moved:"
    )
    print(c)

File: test_dir/cpp/src/tuples/first.cpp
#include "tuples\first.h"
#include "py_list.h"
#include "py_str.h"
#include "py_tuple.h"
#include "pypp_util/print.h"
#include "pypp_util/to_py_str.h"
#include <any>

void _inline_tuple(PyTup<double, PyStr> tup) { print(tup); }

PyTup<int, double> _get_tup() { return PyTup(1, 2.0); }

void _argument_unpacking(int a, double b) { print(a, b); }

void _arg_unpacking_fail(int a, int b, int c) { print(a, b, c); }

void tuples_fn() {
    print(PyStr("TUPLE RESULTS:"));
    PyTup<int, double, PyStr> a = PyTup(1, 1.2, PyStr("a"));
    print(to_pystr(a.count(2)));
    print(to_pystr(a.index(1.2)));
    int b = a.get<0>();
    print(to_pystr(b));
    print(to_pystr(PyTup(1, 2) == PyTup(1, 2)));
    print(to_pystr(PyTup(1, 2) != PyTup(1, 2)));
    print(to_pystr(PyTup(1, 2) < PyTup(1, 2)));
    print(to_pystr(PyTup(1, 2) <= PyTup(1, 2)));
    print(to_pystr(PyTup(1, 2) > PyTup(1, 2)));
    print(to_pystr(PyTup(1, 2) >= PyTup(1, 2)));
    print(PyTup(1, 2));
    print(PyTup(1, 2, PyStr("a")));
    print(to_pystr(PyTup(1, 2).len()));
    _inline_tuple(PyTup(1.2, PyStr("z")));
    auto [x, y, z] = a;
    print(x, y, z);
    auto [u, v] = _get_tup();
    print(u, v);
    std::apply(_argument_unpacking, _get_tup().raw());
    PyList<int> c = PyList({1, 2, 3});
    PyTup<int, PyList<int>> d = PyTup(1, std::move(c));
    print(d);
    print(PyStr("below will be [1, 2, 3] for Python, but [] for C++ because "
                "the list was moved:"));
    print(c);
}
